polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -126,6 +133,227 @@ void createLocalBarrier(OpBuilder builder) {
+  void addMemRefAccess(affine::MemRefAccess access);
+
+  /// Return true iff no memref accesses in \p accesses are stores.
+  bool areReadOnly(ArrayRef<affine::MemRefAccess> accesses) const;
+
+  /// Return true iff all memref accesses in \p accesses are stores.
+  bool areWriteOnly(ArrayRef<affine::MemRefAccess> accesses) const;
+
+  /// Return true if memref accesses in \p accesses are a mix of loads and
+  /// stores.
+  bool areReadWrite(ArrayRef<affine::MemRefAccess> accesses) const;
+
+  /// Determine whether the memref \access exhibits temporal reuse.
+  bool hasTemporalReuse(const affine::MemRefAccess &access,
+                        const SmallVectorImpl<Value> &threadVars) const;
+
+private:
+  const MemoryAccessAnalysis &memAccessAnalysis;
+
+  DataFlowSolver &solver;
+
+  /// Collects all memory accesses for a given memref value.
+  DenseMap<Value, SmallVector<affine::MemRefAccess>> accesses;
+
+  /// The preferred memory space for each memref access;
+  DenseMap<const affine::MemRefAccess *, MemorySpace> accessToMemSpace;
+};
+
+std::optional<MemorySelector::MemorySpace>
+MemorySelector::selectMemorySpace(Value memref) const {
+  assert(isa<MemRefType>(memref.getType()) && "Expecting a memref");
+
+  auto it = accesses.find(memref);
+  if (it == accesses.end())
+    return std::nullopt;
+
+  auto numShared = [this](ArrayRef<affine::MemRefAccess> accesses) {
+    return llvm::count_if(accesses, [this](affine::MemRefAccess access) {
+      auto it = accessToMemSpace.find(&access);
+      if (it == accessToMemSpace.end())
+        return false;
+      return (it->second == MemorySpace::Shared);
+    });
+  };
+
+  /// Recommend shared memory if at least half of the accesses for this memref
+  /// should use shared memory.
+  ArrayRef<affine::MemRefAccess> accesses = it->second;
+  if (numShared(accesses) >= std::ceil((double)accesses.size() / 2))
+    return MemorySpace::Shared;
+
+  return MemorySpace::Global;
+}
+
+void MemorySelector::analyze(LoopLikeOpInterface loop) {
+  assert(accesses.empty() && accessToMemSpace.empty() &&
+         "Expecting empty maps");
+
+  // Collect the global thread ids used in the function the loop is in.
+  auto funcOp = loop->template getParentOfType<FunctionOpInterface>();
+  SmallVector<Value> threadVars =
+      memAccessAnalysis.computeThreadVector(funcOp, solver);
+
+  // Collect candidate memref accesses in the loop.
+  loop->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (!isa<affine::AffineLoadOp, affine::AffineStoreOp>(op))
+      return;
+
+    affine::MemRefAccess access(op);
+    addMemRefAccess(access);
+  });
+
+  // Analyze the accesses collected and populate the 'accessToMemSpace' map.
+  for (auto &entry : accesses) {
+    ArrayRef<affine::MemRefAccess> accesses = entry.second;
+
+    // Filter out memRef accesses that aren't read-only.
+    if (CollectReadOnlyAccesses && !areReadOnly(accesses))
+      continue;
+
+    for (const affine::MemRefAccess &access : accesses) {
```

Comment:
With more thoughts, we probably don't need this loop and `accessToMemSpace`, as all `access` in `accesses` have the same `memref`, and their memory access matrix should be the same.

Reply:
We'll need to think about this and address in a future PR (there are 2 that depend on this one and we need more to finish the feature). I'll push this in and address the comment after the current set of dependent PRs are merged in.